keras_cv/applications/object_detection/retina_net/__internal__/utils/anchor_box.py

```diff
@@ -0,0 +1,123 @@
+        self.aspect_ratios = aspect_ratios or [0.5, 1.0, 2.0]
+        self.scales = scales or [2**x for x in [0, 1 / 3, 2 / 3]]
+
+        self._num_anchors = len(self.aspect_ratios) * len(self.scales)
+        self._strides = strides or [2**i for i in range(3, 8)]
+        self._areas = areas or [x**2 for x in [32.0, 64.0, 128.0, 256.0, 512.0]]
+        self._anchor_dims = self._compute_dims()
+
+    def _compute_dims(self):
+        """Computes anchor box dimensions for all ratios and scales at all levels
+        of the feature pyramid.
+        """
+        anchor_dims_all = []
+        for area in self._areas:
+            anchor_dims = []
+            for ratio in self.aspect_ratios:
+                anchor_height = tf.math.sqrt(area / ratio)
+                anchor_width = area / anchor_height
+                dims = tf.reshape(
+                    tf.stack([anchor_width, anchor_height], axis=-1), [1, 1, 2]
+                )
+                for scale in self.scales:
+                    anchor_dims.append(scale * dims)
+            anchor_dims_all.append(tf.stack(anchor_dims, axis=-2))
+        return anchor_dims_all
+
+    def _get_anchors(self, feature_height, feature_width, level):
+        """Generates anchor boxes for a given feature map size and level
+        Arguments:
+          feature_height: An integer representing the height of the feature map.
+          feature_width: An integer representing the width of the feature map.
+          level: An integer representing the level of the feature map in the
+            feature pyramid.
+        Returns:
+          anchor boxes with the shape
+          `(feature_height * feature_width * num_anchors, 4)`
+        """
+        rx = tf.range(feature_width, dtype=tf.float32) + 0.5
+        ry = tf.range(feature_height, dtype=tf.float32) + 0.5
+        centers = tf.stack(tf.meshgrid(rx, ry), axis=-1) * self._strides[level - 3]
+        centers = tf.expand_dims(centers, axis=-2)
+        centers = tf.tile(centers, [1, 1, self._num_anchors, 1])
+        dims = tf.tile(
+            self._anchor_dims[level - 3], [feature_height, feature_width, 1, 1]
+        )
+        anchors = tf.concat([centers, dims], axis=-1)
+        return tf.reshape(
+            anchors, [feature_height * feature_width * self._num_anchors, 4]
+        )
+
+    def call(self, image_height, image_width):
+        """Generates anchor boxes for all the feature maps of the feature pyramid.
+        Arguments:
+          image_height: Height of the input image.
+          image_width: Width of the input image.
+        Returns:
+          anchor boxes for all the feature maps, stacked as a single tensor
+            with shape `(total_anchors, 4)`
+        """
+        anchors = [
+            self._get_anchors(
+                tf.math.ceil(image_height / 2**i),
+                tf.math.ceil(image_width / 2**i),
+                i,
+            )
+            for i in range(3, 8)
```

Comment:
The magic number here will require some comment.

Reply:
Lets fine-tune this in a follow up PR instead of investing tons of time into documenting this implementation.  We'll be replacing this with Zhenyu's old AnchorBox generator that MG uses